File: frontend/src/api/notes.ts
/**
 * @module api/notes
 * @description Note CRUD and action API endpoints. Notes belong to notebooks
 * and support pinning, archiving, and tag attachment/detachment.
 * 笔记CRUD和操作API端点。笔记属于笔记本，支持置顶、归档和标签的添加/移除。
 */

import client from './client'
import type { Note, NoteCreateRequest, NoteUpdateRequest } from '@/types/note'
import type { PaginatedResponse } from '@/types/common'

/** Note API methods.
 * 笔记API方法。
 */
export const notesApi = {
  /**
   * List all notes within a specific notebook.
   * 列出特定笔记本内的所有笔记。
   *
   * @param notebookId - Parent notebook ID.
   *                     父笔记本ID。
   * @param params     - Optional query parameters.
   *                     可选查询参数。
   * @param params.pinned   - Filter by pinned status.
   *                          按置顶状态过滤。
   * @param params.archived - Filter by archived status.
   *                          按归档状态过滤。
   * @param params.tag_id   - Filter by tag ID.
   *                          按标签ID过滤。
   * @param params.page     - Page number (1-based).
   *                          页码（从1开始）。
   * @param params.size     - Items per page.
   *                          每页项目数。
   * @returns Paginated list of notes.
   *          分页的笔记列表。
   */
  getAll(
    notebookId: string,
    params?: { pinned?: boolean; archived?: boolean; tag_id?: string; page?: number; size?: number }
  ): Promise<PaginatedResponse<Note>> {
    return client.get(`/notebooks/${notebookId}/notes`, { params }).then((r) => r.data)
  },

  /**
   * Fetch a single note by its ID.
   * 根据ID获取单个笔记。
   *
   * @param id - The note's unique identifier.
   *             笔记的唯一标识符。
   * @returns The requested note.
   *          请求的笔记。
   */
  getById(id: string): Promise<Note> {
    return client.get(`/notes/${id}`).then((r) => r.data)
  },

  /**
   * Create a new note inside a notebook.
   * 在笔记本内创建新笔记。
   *
   * @param notebookId - Parent notebook ID.
   *                     父笔记本ID。
   * @param data       - Creation payload (title, optional content and tag IDs).
   *                     创建负载（标题、可选内容和标签ID）。
   * @returns The newly created note.
   *          新创建的笔记。
   */
  create(notebookId: string, data: NoteCreateRequest): Promise<Note> {
    return client.post(`/notebooks/${notebookId}/notes`, data).then((r) => r.data)
  },

  /**
   * Update an existing note.
   * 更新现有笔记。
   *
   * @param id   - The note's unique identifier.
   *               笔记的唯一标识符。
   * @param data - Fields to update.
   *               要更新的字段。
   * @returns The updated note.
   *          更新后的笔记。
   */
  update(id: string, data: NoteUpdateRequest): Promise<Note> {
    return client.put(`/notes/${id}`, data).then((r) => r.data)
  },

  /**
   * Delete a note permanently.
   * 永久删除笔记。
   *
   * @param id - The note's unique identifier.
   *             笔记的唯一标识符。
   * @returns Resolves when deletion is complete.
   *          删除完成时解析。
   */
  delete(id: string): Promise<void> {
    return client.delete(`/notes/${id}`)
  },

  /**
   * Toggle the pinned status of a note.
   * 切换笔记的置顶状态。
   *
   * @param id        - The note's unique identifier.
   *                    笔记的唯一标识符。
   * @param isPinned  - `true` to pin, `false` to unpin.
   *                    `true`表示置顶，`false`表示取消置顶。
   * @returns The updated note.
   *          更新后的笔记。
   */
  pin(id: string, isPinned: boolean): Promise<Note> {
    return client.patch(`/notes/${id}/pin`, { is_pinned: isPinned }).then((r) => r.data)
  },

  /**
   * List all notes across every notebook.
   * 列出所有笔记本中的所有笔记。
   *
   * @param params - Optional query parameters.
   *                 可选查询参数。
   * @param params.page   - Page number (1-based).
   *                        页码（从1开始）。
   * @param params.size   - Items per page.
   *                        每页项目数。
   * @param params.tag_id - Filter by tag ID.
   *                        按标签ID过滤。
   * @returns Paginated list of notes.
   *          分页的笔记列表。
   */
  getAllNotes(params?: { page?: number; size?: number; tag_id?: string }): Promise<PaginatedResponse<Note>> {
    return client.get('/notes', { params }).then((r) => r.data)
  },

  /**
   * List all archived notes across every notebook.
   * 列出所有笔记本中的所有已归档笔记。
   *
   * @param params - Optional query parameters.
   *                 可选查询参数。
   * @param params.page - Page number (1-based).
   *                      页码（从1开始）。
   * @param params.size - Items per page.
   *                      每页项目数。
   * @returns Paginated list of archived notes.
   *          分页的已归档笔记列表。
   */
  getArchivedNotes(params?: { page?: number; size?: number }): Promise<PaginatedResponse<Note>> {
    return client.get('/notes/archived', { params }).then((r) => r.data)
  },

  /**
   * Toggle the archived status of a note.
   * 切换笔记的归档状态。
   *
   * @param id          - The note's unique identifier.
   *                      笔记的唯一标识符。
   * @param isArchived  - `true` to archive, `false` to unarchive.
   *                      `true`表示归档，`false`表示取消归档。
   * @returns The updated note.
   *          更新后的笔记。
   */
  archive(id: string, isArchived: boolean): Promise<Note> {
    return client.patch(`/notes/${id}/archive`, { is_archived: isArchived }).then((r) => r.data)
  },

  /**
   * Attach a tag to a note.
   * 为笔记添加标签。
   *
   * @param noteId - The note's unique identifier.
   *                 笔记的唯一标识符。
   * @param tagId  - The tag's unique identifier.
   *                 标签的唯一标识符。
   * @returns The updated note with the new tag attached.
   *          添加了新标签的更新后的笔记。
   */
  attachTag(noteId: string, tagId: string): Promise<Note> {
    return client.post(`/notes/${noteId}/tags`, { tag_id: tagId }).then((r) => r.data)
  },

  /**
   * Detach a tag from a note.
   * 从笔记移除标签。
   *
   * @param noteId - The note's unique identifier.
   *                 笔记的唯一标识符。
   * @param tagId  - The tag's unique identifier.
   *                 标签的唯一标识符。
   * @returns The updated note with the tag removed.
   *          移除了标签的更新后的笔记。
   */
  detachTag(noteId: string, tagId: string): Promise<Note> {
    return client.delete(`/notes/${noteId}/tags/${tagId}`).then((r) => r.data)
  },
}

File: backend/app/api/v1/notes.py
"""Note management endpoints.

Provides CRUD operations for notes, including listing, creating,
updating, deleting, pinning, archiving, and tag attachment/detachment.
Notes are scoped to notebooks but can also be queried globally.
"""

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.common import PaginatedResponse
from app.schemas.note import (
    NoteArchiveUpdate,
    NoteCreate,
    NotePinUpdate,
    NoteResponse,
    NoteTagAttach,
    NoteUpdate,
)
from app.services.note import note_service

router = APIRouter()

# Static routes must come before parameterized routes to avoid
# FastAPI interpreting "archived" or "notes" as a {note_id} path parameter.


@router.get("/notes", response_model=PaginatedResponse[NoteResponse])
async def list_all_notes(
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    tag_id: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List all notes across all notebooks, optionally filtered by tag.

    Args:
        page: Page number (1-indexed).
        size: Number of items per page (max 100).
        tag_id: Optional tag UUID to filter notes by.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        PaginatedResponse[NoteResponse]: Paginated list of notes.
    """
    items, total = await note_service.list_all_notes(
        db, current_user.id, page=page, size=size, tag_id=tag_id,
    )
    # Ceiling division to compute total page count
    pages = (total + size - 1) // size
    return PaginatedResponse(items=items, total=total, page=page, size=size, pages=pages)


@router.get("/notes/archived", response_model=PaginatedResponse[NoteResponse])
async def list_archived_notes(
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List all archived notes.

    Args:
        page: Page number (1-indexed).
        size: Number of items per page (max 100).
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        PaginatedResponse[NoteResponse]: Paginated list of archived notes.
    """
    items, total = await note_service.list_archived_notes(db, current_user.id, page=page, size=size)
    pages = (total + size - 1) // size
    return PaginatedResponse(items=items, total=total, page=page, size=size, pages=pages)


@router.get("/notebooks/{notebook_id}/notes", response_model=PaginatedResponse[NoteResponse])
async def list_notes(
    notebook_id: uuid.UUID,
    pinned: bool | None = Query(None),
    archived: bool = Query(False),
    tag_id: str | None = Query(None),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List notes within a specific notebook with optional filters.

    Args:
        notebook_id: UUID of the parent notebook.
        pinned: Optional filter for pinned status (None = all).
        archived: Whether to include archived notes. Defaults to False.
        tag_id: Optional tag UUID to filter notes by.
        page: Page number (1-indexed).
        size: Number of items per page (max 100).
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        PaginatedResponse[NoteResponse]: Paginated list of notes.

    Raises:
        NotFoundException: If the notebook does not exist.
        ForbiddenException: If the notebook does not belong to the current user.
    """
    items, total = await note_service.list_notes(
        db, notebook_id, current_user.id, pinned=pinned, archived=archived, tag_id=tag_id, page=page, size=size
    )
    pages = (total + size - 1) // size
    return PaginatedResponse(items=items, total=total, page=page, size=size, pages=pages)


@router.post("/notebooks/{notebook_id}/notes", response_model=NoteResponse, status_code=201)
async def create_note(
    notebook_id: uuid.UUID,
    data: NoteCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new note within a notebook.

    Args:
        notebook_id: UUID of the parent notebook.
        data: Note creation payload (title, content, etc.).
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The newly created note.

    Raises:
        NotFoundException: If the notebook does not exist.
        ForbiddenException: If the notebook does not belong to the current user.
    """
    return await note_service.create(db, notebook_id, current_user.id, data)


@router.get("/notes/{note_id}", response_model=NoteResponse)
async def get_note(
    note_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Retrieve a single note by ID.

    Args:
        note_id: UUID of the note to retrieve.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The requested note.

    Raises:
        NotFoundException: If the note does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.get_note(db, note_id, current_user.id)


@router.put("/notes/{note_id}", response_model=NoteResponse)
async def update_note(
    note_id: uuid.UUID,
    data: NoteUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update an existing note.

    Args:
        note_id: UUID of the note to update.
        data: Partial update payload with fields to change.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The updated note.

    Raises:
        NotFoundException: If the note does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.update(db, note_id, current_user.id, data)


@router.delete("/notes/{note_id}", status_code=204)
async def delete_note(
    note_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete a note by ID.

    Args:
        note_id: UUID of the note to delete.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        None: 204 No Content on success.

    Raises:
        NotFoundException: If the note does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    await note_service.delete(db, note_id, current_user.id)


@router.patch("/notes/{note_id}/pin", response_model=NoteResponse)
async def pin_note(
    note_id: uuid.UUID,
    data: NotePinUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Toggle the pinned status of a note.

    Args:
        note_id: UUID of the note to pin/unpin.
        data: Payload containing the pinned flag.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The updated note with new pinned status.

    Raises:
        NotFoundException: If the note does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.pin(db, note_id, current_user.id, data)


@router.patch("/notes/{note_id}/archive", response_model=NoteResponse)
async def archive_note(
    note_id: uuid.UUID,
    data: NoteArchiveUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Toggle the archived status of a note.

    Args:
        note_id: UUID of the note to archive/unarchive.
        data: Payload containing the archived flag.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The updated note with new archived status.

    Raises:
        NotFoundException: If the note does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.archive(db, note_id, current_user.id, data)


@router.post("/notes/{note_id}/tags", response_model=NoteResponse)
async def attach_tag(
    note_id: uuid.UUID,
    data: NoteTagAttach,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Attach a tag to a note.

    Args:
        note_id: UUID of the note.
        data: Payload containing the tag_id to attach.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The updated note with the tag attached.

    Raises:
        NotFoundException: If the note or tag does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.attach_tag(db, note_id, current_user.id, data.tag_id)


@router.delete("/notes/{note_id}/tags/{tag_id}", response_model=NoteResponse)
async def detach_tag(
    note_id: uuid.UUID,
    tag_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Detach a tag from a note.

    Args:
        note_id: UUID of the note.
        tag_id: UUID of the tag to detach.
        db: Async database session injected via dependency.
        current_user: Authenticated user resolved from the Bearer token.

    Returns:
        NoteResponse: The updated note with the tag detached.

    Raises:
        NotFoundException: If the note or tag association does not exist.
        ForbiddenException: If the note does not belong to the current user.
    """
    return await note_service.detach_tag(db, note_id, current_user.id, tag_id)
